docs/language/interfaces.mdx

```diff
@@ -544,3 +544,413 @@ resource ExampleToken: FungibleToken {
+```cadence
+pub contract interface Token {
+    pub struct Foo {}
+}
+
+pub contract interface NonFungibleToken: Token {
+    pub struct Foo {}
+}
+
+pub contract MyToken: NonFungibleToken {
+    pub fun test() {
+        let foo: Foo  // This will refer to the `NonFungibleToken.Foo`
+    }
+}
+```
+
+If a user needed to access the `Foo` struct coming from the super interface `Token`, then they can
+access it using the fully qualified name. e.g: `let foo: Token.Foo`.
+
+However, it is not allowed to have two or more inherited type/events definitions with identical names for an interface.
+
+```cadence
+pub contract interface Token {
+    pub struct Foo {}
+}
+
+pub contract interface Collectible {
+    pub struct Foo {}
+}
+
+// Invalid: Two type definitions with the same name from two interfaces.
+pub contract NonFungibleToken: Token, Collectible {
+}
+```
+Similar to default functions, there can be situations where the same type/event definition can be available
+via different inheritance paths.
+
+```cadence
+pub contract interface Logger {
+    pub struct Foo {}
+}
+
+pub contract interface Token: Logger {}
+
+pub contract interface Collectible: Logger {}
+
+// Valid: `Logger.Foo` struct is visible to the `NonFungibleToken` interface via both `Token` and `Collectible`.
+pub contract interface NonFungibleToken: Token, Collectible {}
+```
+
+In the above example, `Logger.Foo` type definition is visible to the `NonFungibleToken` interface via both `Token`
+and `Collectible`. Even though it is available from two different interfaces, they are both referring to the same
+type definition. Therefore, the above code is valid.
+
+However, if at least one of the interfaces in the middle of the chain also overrides the type definition `Foo`,
+then the code becomes invalid, as there are multiple implementations present now, which leads to ambiguity.
+
+```cadence
+pub contract interface Logger {
+    pub struct Foo {}
+}
+
+pub contract interface Token: Logger {
+    pub struct Foo {}
+}
+
+pub contract interface Collectible: Logger {}
+
+// Invalid: The default implementation of the `Foo` struct by the `Logger`
+// interface is visible to the `NonFungibleToken` via the `Collectible` interface.
+// Another implementation of `Foo` struct is visible to the `NonFungibleToken` via the `Token` interface.
+// This creates ambiguity.
+pub resource interface NonFungibleToken: Token, Provider {}
+```
+
+
+### Linearizing Conditions
+
+As mentioned in the [functions with conditions](#functions-with-conditions) section, it would be required to linearize
+the function conditions, to determine the order in which pre- and post-conditions are executed.
+This is done by linearizing the interfaces, and hence conditions, in a **depth-first pre-ordered manner, without duplicates**.
+
+For example, consider an interface inheritance hierarchy as below:
+```
+       A
+      / \
+     B   C
+    / \ /
+   D   E
+
+where an edge from A (top) to B (bottom) means A inherits B.
```

Comment:
This may seem a bit odd to express inheritance hierarchy in a downward manner (i.e.: ancestors at the bottom, and descendants at the top). But it was written this way because it converts to the source code better, and makes more sense when explaining the linearization/traversak below.